Generators subject to constraints

TimSheard opened the pull request:
# Description
Added a new directory   cardano-ledger-test/src/Test/Cardano/Ledger/Constrained
Removed a few Mock constraints from the old random generators
Added some new PrettyC instances to Test.Cardano.Ledger.Generic.PrettyCore 

The point of all this is to build a domain-specific embedded language that can be used to generate random values of a particular type, subject to constraints.

In the Example file is a set of 76 constraints that build a random NewEpochState with all the properties I believe that the real NewEpochState have. Goal is to further refine this, and extend the work to generating random blocks that can be 'applied' to this state, to test for errors.


# Checklist

- [x] Commit sequence broadly makes sense
- [x] Commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] Any changes are noted in the [changelog](https://github.com/input-output-hk/cardano-ledger/blob/master/CHANGELOG.md)
- [x] Code is formatted with fourmolu (which can be run with `scripts/fourmolise.sh`
- [x] Self-reviewed the diff

TimSheard committed:
Generating random values subject to constraints.
Added Data.Universe, that introduces Singleton types
Added TypeRep.hs, a closed Singleton for use in typed terms
Added data types Term Pred Literal that use Singletons to track types.
Added the Typed monad to track Pred failures
Added Spec and functions (Pred -> Typed Spec) and (Spec -> Spec -> Spec)
Added Generators that (Spec t -> Typed (Gen t))
Added Lenses for types inside NewEpochState
Added predicate Variables for every component of NewEpochState
Added Env (generated by the constraints) that can be concretized to NewEpochState
Removed some (Mock c) constraints from generators in Shelley, Alonzo, Babbage Ers.

UlfNorell committed:
Start on QuickCheck properties for constraint language

TimSheard committed:
Added the Size type, Property tests for each Spec type.
Added new style RelSpec (SumV) which uses Size as Range.
Moved Size and SumV to own file. SumCond renamed as OrdCond.
Adds class methods adjust, partitionBy, greater, were removed.
Cleanup, restructured Monad(no dependency)

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,707 @@
+negSumCond EQL = EQL
+negSumCond LTH = GTH
+negSumCond LTE = GTH
+negSumCond GTH = LTH
```
Shouldn't negation be like this: `not (x < y) == x >= y` and `not (x > y) == x <= y`?
```suggestion
negSumCond LTH = GTE
negSumCond LTE = GTH
negSumCond GTH = LTE
```

TimSheard replied:
This is more subtle than that, because the domain is just the positive numbers, and not every number has a predecessor. So this transformation is sound, but not complete.

lehins replied:
That sounds good. Can you leave a short comment about it for the future reader? Cause it looks wrong, and someone in the future might try to wrongfully "fix" it.

TimSheard replied:
Ok, we now call it reverseOrdCond, give an example of why it is needed, and give a ink to a URL where they explain how it arises.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Monad.hs:
```diff
@@ -0,0 +1,181 @@
+
+-- | Pushes the (Left msgs) into a call to 'error', then injects into a Monad
+monadTyped :: Monad m => Typed t -> m t
+monadTyped t = pure (errorTyped t)
```
It might be better to make it strict, otherwise it is possible to carry a thunk with bottom around, unless that is desired, of course.
```suggestion
monadTyped t = pure $! errorTyped t
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs:
```diff
@@ -0,0 +1,1324 @@
+  spec1 <- genSize
+  spec2 <- genSize
+  case (spec1 <> spec2) of
+    SzNever _xs -> pure True -- trace (unlines _xs) $ pure True
```
Dead code
```suggestion
    SzNever _xs -> pure True
```

TimSheard replied:
done

lehins commented on eras/alonzo/test-suite/src/Test/Cardano/Ledger/Alonzo/Serialisation/Generators.hs:
```diff
@@ -87,7 +87,8 @@ import qualified PlutusLedgerApi.V1 as PV1
 import Test.Cardano.Ledger.Binary.Twiddle
-import Test.Cardano.Ledger.Shelley.ConcreteCryptoTypes (Mock)
+
+-- import Test.Cardano.Ledger.Shelley.ConcreteCryptoTypes (Mock)
```
Dead code
```suggestion
```

TimSheard replied:
done

lehins commented on libs/cardano-data/src/Data/Universe.hs:
```diff
@@ -0,0 +1,202 @@
+-- Each one adding more detail, so that the types can do more
+
+-- | Hide the index for any indexed type 't'
+data Any t where Any :: (t i) -> Any t
```
It is a bit hard to read type definitions with `where` in the middle. Also redundant parenthesis. I suggest:
```suggestion
data Any t where
  Any :: t i -> Any t
```

TimSheard replied:
done

lehins commented on libs/cardano-data/src/Data/Universe.hs:
```diff
@@ -0,0 +1,202 @@
+deriving instance (forall i. Show (t i)) => Show (Any t)
+
+-- | Hide the index for a singleton type 't'
+data Some t where Some :: Singleton t => (t i) -> Some t
```
Same comment as above
```suggestion
data Some t where
  Some :: Singleton t => t i -> Some t
```

TimSheard replied:
done

lehins commented on libs/cardano-data/src/Data/Universe.hs:
```diff
@@ -0,0 +1,202 @@
+-- | Hide the index for a singleton type 't'
+data Some t where Some :: Singleton t => (t i) -> Some t
+
+data Dyn rep where Dyn :: Singleton rep => rep t -> t -> Dyn rep
```
This inconsistency is really confusing `t` vs `i` vs `rep`. We should really stick to one meaning for the type variable, whichever it will be (either `rep` or `t`). Also
```suggestion
data Dyn t where
 Dyn :: Singleton t => t i -> i -> Dyn t
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Env.hs:
```diff
@@ -0,0 +1,154 @@
+
+-- We are ignoring the Accessfield on purpose
+
+data Env era = Env (Map String (Payload era))
```
Unless laziness is required, it is better to use `newtype`
```suggestion
newtype Env era = Env (Map String (Payload era))
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Ast.hs:
```diff
@@ -0,0 +1,453 @@
+
+-- | Version of (:$) That takes a Term on the right, rather than a Target
+($>) :: Target era (a -> t) -> Term era a -> Target era t
+($>) f x = f :$ (Simple x)
```
It is better not to use the same names (and especially operators) as the ones from `base` (eg. [`($>)`](https://downloads.haskell.org/ghc/9.2.5/docs/html/libraries/base-4.16.4.0/Data-Functor.html#v:-36--62-)
Also redundant parenthesis
```suggestion
($>) f x = f :$ Simple x
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/cardano-ledger-test.cabal:
```diff
@@ -78,6 +93,7 @@ library
         cardano-protocol-tpraos,
         cardano-slotting,
         containers,
+        hspec-expectations,
```
There is never really a need to import `hspec-expectation` directly, they are all re-exported from `Test.HSpec`
```suggestion
        hspec,
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Ast.hs:
```diff
@@ -0,0 +1,453 @@
+simplifySum (Project _ (Fixed (Lit _ m))) = pure (List.foldl' (\ans x -> add ans (getsum x)) zero m)
+simplifySum x = failT ["Can't simplify Sum: " ++ show x ++ ", to a value."]
+
+-- | Fully evaluate an Term, looking up the variables in the Env.
```
```suggestion
-- | Fully evaluate a `Term`, looking up the variables in the `Env`.
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Ast.hs:
```diff
@@ -0,0 +1,453 @@
+termRep (Var (V _ r _)) = r
+termRep (Dom t) = SetR r
+  where
+    MapR r _ = termRep t
```
Since you already have `ViewPatterns` enabled you can do something like this:
```suggestion
termRep (Dom (termRep -> MapR r _)) = SetR r
```

TimSheard replied:
This is incredibly clever. Thanks.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Ast.hs:
```diff
@@ -0,0 +1,453 @@
+    MapR a _ = termRep x
+termRep (ProjS _ t x) = SetR t
+  where
+    SetR _a = termRep x
```
I am just curious, why this pattern match is needed? Is it a sanity check to ensure that `termRep` produces the correct case `SetR` for the `Term era (Set t)`, or is it actually required to satisfy the type checker?

TimSheard replied:
You are right,it is not needed.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Ast.hs:
```diff
@@ -0,0 +1,453 @@
+makeTest :: Env era -> Pred era -> Typed (String, Bool, Pred era)
+makeTest env c = do
+  b <- runPred env c
+  pure (show c ++ " => " ++ show b, b, c)
```
> `show b, b, c`
Are they gonna show it on BBC? :smile:

TimSheard replied:
of course. All my code is shown on the BBC!

JaredCorduan replied:
:rofl:

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,707 @@
+{-# LANGUAGE TypeSynonymInstances #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- {-# LANGUAGE AllowAmbiguousTypes #-}
```
Dead code
```suggestion
```

TimSheard replied:
Caught this one my self, on my latest self review.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,707 @@
+import qualified Cardano.Ledger.Shelley.PParams as Core (ProposedPPUpdates (..))
+
+-- import  Cardano.Ledger.Conway.Governance(ConwayTallyState(..))
+-- import Lens.Micro
```
Dead code?
```suggestion
```

TimSheard replied:
this one too.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,707 @@
+  adjust :: [String] -> SumCond -> Int -> x -> Gen x
+
+  partitionBy :: [String] -> SumCond -> Int -> x -> Gen [x]
+  partitionBy msgs cond count total = do b <- adjust msgs cond count total; partition msgs count b
```
That's exactly what bind is for:
```suggestion
  partitionBy msgs cond count total = adjust msgs cond count total >>= partition msgs count
```

TimSheard replied:
This function has been removed.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,707 @@
+
+class (Ord x, Show x) => Adds x where
+  add :: x -> x -> x
+  subtract :: [String] -> x -> x -> x
```
I think it is a bad idea to shadow a function from `Prelude`, especially one with very similar in functionality, but with arguments flipped. Maybe use `plus` and `minus` instead of `add` and `subtract`?

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,707 @@
+
+naturalPartition :: [String] -> Natural -> Int -> Natural -> Gen [Natural]
+naturalPartition msgs smallest size total =
+  map (fromIntegral) <$> intPartition msgs "Natural" (fromIntegral smallest) size (fromIntegral total)
```
Hopefully `naturalPartition` is not used with very large numbers, otherwise this thing will overflow and result in a mysterious `Underflow` exception 
```suggestion
  map fromIntegral <$> intPartition msgs "Natural" (fromIntegral smallest) size (fromIntegral total)
```

TimSheard replied:
The only use I can find is in small tests.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,707 @@
+  subtract :: [String] -> x -> x -> x
+  zero :: x
+  partition :: [String] -> Int -> x -> Gen [x]
+  addCount :: x -> Int
```
What does "count" mean?

TimSheard replied:
This has been made obsolete and removed.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,707 @@
+  subtract _ = (-)
+  zero = 0
+  partition = rationalPartition
+  addCount _ = 2 -- Not as arbitrary as it seems
```
This line is just asking for a question :smile: 

Why `2`, i.e. why is it not arbitrary.

TimSheard replied:
This has been removed, as well as addCount is gone.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,707 @@
+              ++ " Cannot be solved."
+          )
+          msgs
+      else choose (count + 1, total - 1) -- (\ n -> count < n && n < total)
```
Dead code or comment for later?
```suggestion
      else choose (count + 1, total - 1)
```

TimSheard replied:
It was supposed to be a comment about a property that the choose was obeying. But is has been replaced, so t is gone.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,707 @@
+              ++ " Cannot be solved."
+          )
+          msgs
+      else choose (count + 1, total) -- (\ n -> count < n && n <= total)
```
Same here
```suggestion
      else choose (count + 1, total)
```

TimSheard replied:
same result, removed

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,707 @@
+-- The Sums class, for summing a projected c (where Adds c) from some richer type
+
+class (Show x, Adds x) => Sums t x | t -> x where
+  getsum :: t -> x
```
Camel case
```suggestion
  getSum :: t -> x
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,707 @@
+    Mary _ -> pure $ TxOut p (txout & txOutCoinL .~ coins)
+    Alonzo _ -> pure $ TxOut p (txout & txOutCoinL .~ coins)
+    Babbage _ -> pure $ TxOut p (txout & txOutCoinL .~ coins)
+    Conway _ -> pure $ TxOut p (txout & txOutCoinL .~ coins)
```
Avoid boiler plate, use camel case and redundant parenthesis
```suggestion
genTxOutX :: Reflect era => Proof era -> Coin -> Gen (TxOut era)
genTxOutX proof coins = do
  TxOut p txOut <- genTxOut proof
  pure $ TxOut p (txOut & txOutCoinL .~ coins)
```

TimSheard replied:
This may be ugly, but it prevents polluting the types of about 100 functions with the Core.EraTxOut era constraint.
I remove the parenthesis.

lehins replied:
Note that I didn't add `EraTxOut`, I added `Reflect` constraint, which implies `EraTxOut`.  You said it yourself it is ok to use `Reflect` in generic tests. Also this is what I mentioned during the meeting today. If we can avoid pattern matching on `Proof` we should, because it will reduce amount of work we have to do during addition of each era.

TimSheard replied:
But, it does not supply the Arbitrary instance for the type family TxOut. 
We need (Arbitrary (TxOut era)), and the pattern match supplies that.

lehins replied:
Just give it a try and you will see that I am right. This is the part that takes care of `Arbitrary`:
```
TxOut p txOut <- genTxOut proof
```

TimSheard replied:
OK only one use of the lens. done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,707 @@
+  partition :: [String] -> Int -> x -> Gen [x]
+  addCount :: x -> Int
+  fromCount :: Int -> x
+  greater :: Int -> Gen x -- Generate an 'x' larger than 'Int'
```
That's better as a haddock
```suggestion
 -- | Generate an 'x' larger than 'Int'
  greater :: Int -> Gen x
```

TimSheard replied:
The greater method has been removed.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,707 @@
+  add = (+)
+  subtract _ = (-)
+  zero = 0
+  partition msgs count total = map fromIntegral <$> (intPartition msgs "Int" 1 count (fromIntegral total))
```
```suggestion
  partition msgs count total = map fromIntegral <$> intPartition msgs "Int" 1 count (fromIntegral total)
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,707 @@
+            ++ typname
+        )
+        msgs
+  | size < 1 =
```
To emphasize that `0` is a totally separate case with a special error message
```suggestion
  | size < 0 =
```

TimSheard replied:
I changed ti it   'size <= 0'

lehins replied:
:smile: But that's the same thing as `< 1`

My comment was about the fact that the `==0` case will never be triggered here because of the clause earlier

TimSheard replied:
This has been fixed

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,707 @@
+            ++ show smallest
+        )
+        msgs
+  | smallest < 0 = errorMess ("The minimum choice must be positive : " ++ show smallest) msgs
```
So, smallest can be `0`? If that's the case:
```suggestion
  | smallest < 0 = errorMess ("The minimum choice must be non-negative: " ++ show smallest) msgs
```
otherwise
```suggestion
  | smallest <= 0 = errorMess ("The minimum choice must be positive: " ++ show smallest) msgs
```

TimSheard replied:
So in the new algorithm, smallest can be any number. This allows for summing DeltaCoin where some of them are negative. For most other types the choices are 0 or 1.  In the new system, the user gets to pick what the smallest number is. So all tests about smallest have been removed. So a constraint looks like

(SumsTo smallest x <= [a,b,c])  denoting we must generate random values for x,a,b,c such that 
x >= smallest, a >= smallest, b >= smallest, c >= smallest and x = a + b + c

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,707 @@
+  let iScale = n * 1000
+      rScale :: Rational
+      rScale = fromIntegral iScale
+  is <- intPartition msgs "Rational" (iScale `div` n) n (round (total * rScale))
```
Slight simplification
```suggestion
  let s = 1000
      rScale :: Rational
      rScale = fromIntegral (n * s)
  is <- intPartition msgs "Rational" s n (round (total * rScale))
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,707 @@
+    ++ "Try adding (Sized (Range 1 m) (Dom x)) constraint to force 'x' to have at least 1 element"
+
+rationalPartition :: [String] -> Int -> Rational -> Gen [Rational]
+rationalPartition msgs 0 total = errorMess (zeroCount "rationalPartition" total) msgs
```
This case is redundant with simplification below
```suggestion
```

TimSheard replied:
this has been restuctured and is gone

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,707 @@
+  getsum (TxOut (Mary _) t) = coin (t ^. Core.valueTxOutL)
+  getsum (TxOut (Alonzo _) t) = coin (t ^. Core.valueTxOutL)
+  getsum (TxOut (Babbage _) t) = coin (t ^. Core.valueTxOutL)
+  getsum (TxOut (Conway _) t) = coin (t ^. Core.valueTxOutL)
```
All of the pattern matching is totally redundant:
```suggestion
instance (Reflect era) => Sums (TxOut era) Coin where
  getSum (TxOut _ txOut) = txOut ^. coinTxOutL
```

TimSheard replied:
got it

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,707 @@
+
+import Cardano.Ledger.BaseTypes (EpochNo (..), ProtVer (..))
+import Cardano.Ledger.Coin (Coin (..), DeltaCoin (..))
+import qualified Cardano.Ledger.Core as Core
```
There is no need import `Core` qualified. We took care of all of the name clashes that required us to do so. In fact I continuously remove the `Core` and `CC` qualified import from all other modules
```suggestion
import Cardano.Ledger.Core
```

TimSheard replied:
done

lehins commented on libs/cardano-data/src/Data/Universe.hs:
I suspect this is just an example, right? At some point later we'll probably want to move it into hadock.
```suggestion
-- A very simple closed Singleton type
```

TimSheard replied:
yes just an example,

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Env.hs:
```diff
@@ -0,0 +1,154 @@
+-- ============================================
+-- Group a bunch of bindings into a list
+
+data P era where P :: V era t -> t -> P era
```
Maybe it's me, but it's weird to see GADTs defined like this
```suggestion
data P era where
  P :: V era t -> t -> P era
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Examples.hs:
```diff
@@ -0,0 +1,745 @@
+
+genMaybeCounterExample :: Proof era -> String -> Bool -> OrderInfo -> [Pred era] -> Assembler era -> Gen (Maybe String)
+genMaybeCounterExample proof _testname loud order cs target = do
+  let cs2 = (remDom cs)
```
```suggestion
  let cs2 = remDom cs
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Open.hs:
```diff
@@ -0,0 +1,140 @@
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE ViewPatterns #-}
+
+module Test.Cardano.Ledger.Constrained.Open where
```
This PR is already huge, could you remove this module, since it is not being used.

TimSheard replied:
OK

TimSheard replied:
Yes it is removed.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,707 @@
+
+-- ======
+data PParams era where
+  PParams :: Proof era -> Core.PParams era -> PParams era
```
I tried so hard to get rid of all the name clashes we had before. Could we please not introduce any more new ones.

The goal is that in any module in this repository we should be able to import `Cardano.Ledger.Core` unqualified. This is necessary because there should never be an ambiguity with any of our core types.

TimSheard replied:
done,  we call them PParamsF,  TxOutF  because they are type "F"amilies

lehins replied:
"F"antastic! :smiley:

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,707 @@
+data ProposedPPUpdates era where
+  ProposedPPUpdates :: Proof era -> Core.ProposedPPUpdates era -> ProposedPPUpdates era
+
+-- newtype Core.ProposedPPUpdates era = Core.ProposedPPUpdates (Map (KeyHash 'Genesis (EraCrypto era)) (Core.PParamsUpdate era))
```
Dead code
```suggestion
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Env.hs:
```diff
@@ -0,0 +1,154 @@
+-- V
+
+-- | A proto variable. May or may not contain a Lens (encoded as Access)
+data V era t where V :: String -> Rep era t -> Access era s t -> V era t
```
```suggestion
data V era t where
  V :: String -> Rep era t -> Access era s t -> V era t
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Env.hs:
```diff
@@ -0,0 +1,154 @@
+
+-- | An existentially quantified (V era t), hiding the 't'
+--   Usefull because unlike (V era t), it has both Eq and Ord instances
+data Name era where Name :: V era t -> Name era
```
```suggestion
data Name era where
  Name :: V era t -> Name era
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Env.hs:
```diff
@@ -0,0 +1,154 @@
+      else case compare n1 n2 of
+        LT -> LT
+        GT -> GT
+        EQ -> compareRep rep1 rep2
```
```suggestion
      else compare n1 n2 <> compareRep rep1 rep2
```

TimSheard replied:
That is actually pretty clever. I had to go look at the monoid instance for Ordering, to understand why it works.
It works because  (<>) on Ordering is left-biased (except for EQ).

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Env.hs:
```diff
@@ -0,0 +1,154 @@
+data V era t where V :: String -> Rep era t -> Access era s t -> V era t
+
+data Access era s t where
+  Yes :: forall s t era. Rep era s -> Lens' s t -> Access era s t
```
Any specific reason for this `forall`? I thought it was to adjust the order of type arguments during `TypeApplication` syntax using `@`, eg.`Yes @(NewEpochState era)`, but I couldn't find any uses cases of `@` with `Yes`.
```suggestion
  Yes :: Rep era s -> Lens' s t -> Access era s t
```
In any case, if the `forall` is indeed necessary, it would be good to add it to `No` as well for consistency and a comment explaining why this deviation form default order is needed

TimSheard replied:
In previous version the order of the type variables was important. But that has changed.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Monad.hs:
```diff
@@ -0,0 +1,181 @@
+
+-- | A Pair of a (Rep era t), and a 't' . Useful to encapsulate 't' in a way
+--   one can interogate its type at runtime.
+data Dyn era where Dyn :: Rep era t -> t -> Dyn era
```
```suggestion
data Dyn era where
  Dyn :: Rep era t -> t -> Dyn era
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Monad.hs:
```diff
@@ -0,0 +1,181 @@
+-- ========================================================
+
+-- | A Pair of a (Rep era t), and a 't' . Useful to encapsulate 't' in a way
+--   one can interogate its type at runtime.
```
```suggestion
--   one can interrogate its type at runtime.
```

TimSheard replied:
this type was removed, as it was not used anywhere.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Monad.hs:
```diff
@@ -0,0 +1,181 @@
+-- ================================================================
+-- Computing runtime evidence of a constraint.
+
+data Id x = Id x
```
Does it require boxing/laziness to work? Otherwise using a `newtype` would make more sense:
```suggestion
newtype Id x = Id x
```
Or simply importing [`Identity`](https://downloads.haskell.org/~ghc/8.10.7/docs/html/libraries/base-4.14.3.0/Data-Functor-Identity.html#t:Identity) from base. Sure, it is a longer name, but at least it is already predefined.

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Monad.hs:
```diff
@@ -0,0 +1,181 @@
+--   't' one can always use (Id t) instead. Eg
+--   With (Id x) <- hasOrd repT (Id t)
+data HasCond c t where
+  With :: c t => (s t) -> HasCond c (s t)
```
Would you like to make it strict?
```suggestion
  With :: c t => !(s t) -> HasCond c (s t)
```
Or remove parenthesis:
```suggestion
  With :: c t => s t -> HasCond c (s t)
```

TimSheard replied:
I don't think strictness is wanted. Often the only thing one wants is the withness of the class.
I removed the ( )

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs:
```diff
@@ -0,0 +1,1324 @@
+  (:~:) (Refl),
+ )
+import Test.Cardano.Ledger.Core.Arbitrary ()
+import Test.Cardano.Ledger.Generic.Proof (C_Crypto, MaryEra)
```
Why do we need to use `C_Crypto` here? We should be able to stick to `StandardCrypto` in almost all of our tests

TimSheard replied:
We now use standard Crypto.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs:
```diff
@@ -0,0 +1,1324 @@
+--   The information stored is used to compute the sets. Some times it is about
+--   what is in the set, and sometimes it is about what is NOT in the set.
+data RelSpec era dom where
+  -- \^ There is no restriction on the domain. Denotes the universe.
```
```suggestion
  -- | There is no restriction on the domain. Denotes the universe.
```

TimSheard replied:
I think the \^  was added by fourmolize. This has appeared a number of times. Is that possible?

lehins replied:
Yes, totally, cause that is invalid syntax and haddock would choke on it. So it is possible that fourmolu fixed it, by escaping the `^` symbol

(Note that it is invalid because there is no type constructor above it, which is what `-- ^` denotes)

lehins reviewed (changes requested):
This is a huge PR. I was able to look just through a portion of it. I'll review more tomorrow.

aniketd commented on libs/cardano-data/src/Data/Universe.hs:
```diff
@@ -0,0 +1,202 @@
+--   A closely related class is (Indexed s t). We would like a total ordering on (t i) and (t j),
+--   where the indices, i and j, are different. We do this by ensuring every value of type (t i)
+--   has a unique Shape, and that the type Shape has a total order. Every closed singleton can
+--   be totally ordered this way. Other indexed type are used to make well type terms. I.e.
```
```suggestion
--   be totally ordered this way. Other indexed types are used to make well-typed terms. I.e.
```

TimSheard replied:
done

aniketd commented on libs/cardano-data/src/Data/Universe.hs:
```diff
@@ -0,0 +1,202 @@
+
+-- ==================================================
+
+-- | Type synonym, so when can use ( :~: ) without TypeOperators
```
```suggestion
-- | Type synonym, so we can use ( :~: ) without TypeOperators
```

TimSheard replied:
done

aniketd commented on libs/cardano-data/src/Data/Universe.hs:
```suggestion
-- A very simple, closed Singleton type
```

TimSheard replied:
done

aniketd commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Ast.hs:
```diff
@@ -0,0 +1,453 @@
+data Pred era where
+  Sized :: Sizeable t => Term era Size -> Term era t -> Pred era
+  (:=:) :: Eq a => Term era a -> Term era a -> Pred era
+  (:<=:) :: Ord a => Term era (Set a) -> Term era (Set a) -> Pred era
```
```suggestion
  (:~=:) :: Ord a => Term era (Set a) -> Term era (Set a) -> Pred era
```
Just a suggestion to avoid confusing this operator with a `<=` comparison-esque connotation. Nevermind if this, or, `(:-=:)` does not look any more useful to you. :slightly_smiling_face:

TimSheard replied:
I am thinking of using this :⊆:   instead

Soupstraw replied:
In a way it is a comparison between sets similar to LTE, since subsets define a partial ordering on the set of all sets, so I don't see a problem with this.

Soupstraw replied:
I'd prefer to not have unicode operators

TimSheard committed:
Addressed the comments and the TODOs of the group reviews
with (tim, alexi,joosep,tedora,aniket) including reverseOrdCond,
simplifying Term and Pred types, more robust combinators (using resize).
Added Haddock comments and used better algorithms as well.
Better rewrite rules, improvements to soundness tests from Quviq.

lehins commented on libs/cardano-data/cardano-data.cabal:
This package needs a minor version bump `1.0.1.0` and a changelog entry that mentioned addition of the new module

TimSheard replied:
Done.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Ast.hs:
```diff
@@ -0,0 +1,423 @@
+      pp (Univ.Any spec) = show spec
+
+args :: Target era t -> [Univ.Any (Target era)]
+args (x :$ xs) = (Univ.Any x) : args xs
```
```suggestion
args (x :$ xs) = Univ.Any x : args xs
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Ast.hs:
```diff
@@ -0,0 +1,423 @@
+  substToEnv more (storeVar v t ans)
+substToEnv ((SubItem _ e) : _) _ = failT ["Not Literal expr in substToEnv: " ++ show e]
+
+data SubItem era where SubItem :: V era t -> Term era t -> SubItem era
```
```suggestion
data SubItem era where
  SubItem :: V era t -> Term era t -> SubItem era
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Ast.hs:
```diff
@@ -0,0 +1,423 @@
+findV :: Subst era -> V era t -> Term era t
+findV [] v@(V _ _ _) = Var v -- If its not in the Subst, return the Var
+findV (SubItem (V n2 rep2 _) kn : more) v@(V n1 rep1 _) =
+  if not (n1 == n2)
```
```suggestion
  if n1 /= n2
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Ast.hs:
```diff
@@ -0,0 +1,423 @@
+substPred :: Subst era -> Pred era -> Pred era
+substPred sub (Sized a b) = Sized (substTerm sub a) (substTerm sub b)
+substPred sub (a :=: b) = (substTerm sub a) :=: (substTerm sub b)
+substPred sub (Subset a b) = (substTerm sub a) `Subset` (substTerm sub b)
```
Great thing about infix operators, they reduce parenthesis :slightly_smiling_face: 
```suggestion
substPred sub (a :=: b) = substTerm sub a :=: substTerm sub b
substPred sub (a `Subset` b) = substTerm sub a `Subset` substTerm sub b
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Ast.hs:
```diff
@@ -0,0 +1,423 @@
+  m <- runTerm env x
+  pure (Set.map (\z -> z ^. l) m)
+runTerm env (Delta x) = do
+  (Coin n) <- runTerm env x
```
```suggestion
  Coin n <- runTerm env x
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Ast.hs:
```diff
@@ -0,0 +1,423 @@
+  (Coin n) <- runTerm env x
+  pure (DeltaCoin n)
+runTerm env (Negate x) = do
+  (DeltaCoin n) <- runTerm env x
```
```suggestion
  DeltaCoin n <- runTerm env x
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Ast.hs:
```diff
@@ -0,0 +1,423 @@
+runTerm _ (Lit _ x) = pure x
+runTerm env (Dom x) = do
+  m <- runTerm env x
+  pure (Map.keysSet m)
```
```suggestion
runTerm env (Dom x) = Map.keysSet <$> runTerm env x
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Ast.hs:
```diff
@@ -0,0 +1,423 @@
+  pure (Map.keysSet m)
+runTerm env (Rng x) = do
+  m <- runTerm env x
+  pure (Set.fromList (Map.elems m))
```
```suggestion
runTerm env (Rng x) = Set.fromList . Map.elems <$> runTerm env x
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,639 @@
+
+-- import Test.Cardano.Ledger.Shelley.Serialisation.EraIndepGenerators ()
+-- import Test.Cardano.Ledger.Shelley.Arbitrary
+-- import Test.Cardano.Ledger.ShelleyMA.Serialisation.Generators ()
```
```suggestion
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,639 @@
+  shuffle,
+  vectorOf,
+ )
+import Prelude hiding (subtract)
```
Hiding no longer necessary
```suggestion
```

TimSheard replied:
done, forgot about this.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,639 @@
+      ms = msgs ++ ["genAdds Int"]
+  fromI _ n = n
+  toI n = n
+  genSmall = elements [-2, 0, 1, 2]
```
No `-1`?

TimSheard replied:
Sure why not.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,639 @@
+    where
+      ms = msgs ++ ["genAdds DeltaCoin"]
+  fromI _ n = DeltaCoin (fromIntegral n)
+  toI (DeltaCoin n) = (fromIntegral n)
```
```suggestion
  toI (DeltaCoin n) = fromIntegral n
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,639 @@
+  genT _ r = IndividualPoolStake r <$> arbitrary
+
+instance (Reflect era) => Sums (TxOutF era) Coin where
+  getSum (TxOutF _ t) = coin (t ^. valueTxOutL)
```
```suggestion
  getSum (TxOutF _ txOut) = txOut ^. coinTxOutL
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,639 @@
+
+genTxOutX :: Reflect era => Proof era -> Coin -> Gen (TxOutF era)
+genTxOutX p coins = do
+  do
```
To many `do`s :smile: 
```suggestion
```

TimSheard replied:
you can never have too many do's. But I removed this one,

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,639 @@
+
+genValueX :: Reflect era => Proof era -> Coin -> Gen (ValueF era)
+genValueX proof cn = do
+  (ValueF p v) <- genValue proof
```
```suggestion
  ValueF p v <- genValue proof
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,639 @@
+
+class Show t => Sizeable t where
+  -- | extract the 'size' of 't'
+  getsize :: t -> Int
```
Maybe camelCase?
```suggestion
  getSize :: t -> Int
```

TimSheard replied:
done, and about 12 renames else where

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,639 @@
+  getsize n = fromIntegral n
+
+instance (Show dom, Show rng) => Sizeable (Map dom rng) where
+  getsize m = (Map.size m)
```
```suggestion
  getsize m = Map.size m
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,639 @@
+  getsize m = (Map.size m)
+
+instance Show t => Sizeable (Set t) where
+  getsize m = (Set.size m)
```
```suggestion
  getsize m = Set.size m
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,639 @@
+  -- | 'canFollow x y', is 'x' an appropriate successor to 'y'
+  canFollow :: t -> t -> Bool
+
+  -- | genrate the predecessor, given the successor
```
```suggestion
  -- | Generate the predecessor
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,639 @@
+  -- | genrate the predecessor, given the successor
+  genPred :: t -> Gen t
+
+  -- | generate the successor, given the predecessor
```
```suggestion
  -- | Generate the successor
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,639 @@
+proposedMapL =
+  lens
+    (\(ProposedPPUpdatesF p x) -> x ^. (proposedCoreL . (coreMapL p)))
+    (\(ProposedPPUpdatesF p x) y -> ProposedPPUpdatesF p (x & (proposedCoreL . (coreMapL p)) .~ y))
```
```suggestion
    (\(ProposedPPUpdatesF p x) -> x ^. (proposedCoreL . coreMapL p))
    (\(ProposedPPUpdatesF p x) y -> ProposedPPUpdatesF p (x & (proposedCoreL . coreMapL p) .~ y))
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,639 @@
+legalCallPartition :: [String] -> String -> Integer -> Int -> Integer -> Maybe [String]
+legalCallPartition msgs typname smallest size total
+  | size == 0 && smallest > 0 -- Just (zeroCount "integerPartition" total : msgs)
+    =
```
Looks like a small piece of forgotten code
```suggestion
  | size == 0 && smallest > 0 =
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,639 @@
+      Just (x : xs) -> errorMess x xs
+      Just [] -> errorMess "legalCallPartition returns []" []
+      Nothing ->
+        let mean = total `div` (fromIntegral (size + 1))
```
```suggestion
        let mean = total `div` fromIntegral (size + 1)
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -0,0 +1,639 @@
+
+partitionNatural :: Natural -> [String] -> Int -> Natural -> Gen [Natural]
+partitionNatural small msgs n total =
+  map fromIntegral <$> integerPartition msgs "Word64" (fromIntegral small) n (fromIntegral total)
```
```suggestion
  map fromIntegral <$> integerPartition msgs "Natural" (fromIntegral small) n (fromIntegral total)
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Combinators.hs:
```diff
@@ -0,0 +1,206 @@
+-- | Report a Gen-time error from the current message 'extra' and the
+--   [messages] 'mess' describing the path to this call site.
+errorMess :: HasCallStack => String -> [String] -> a
+errorMess extra mess = error (unlines ("\nGen-time error" : (reverse (extra : mess))))
```
```suggestion
errorMess extra mess = error (unlines ("\nGen-time error" : reverse (extra : mess)))
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Combinators.hs:
```diff
@@ -0,0 +1,206 @@
+      loop result extra = do
+        i <- choose (0, Set.size extra - 1)
+        loop (Set.insert (Set.elemAt i extra) result) (Set.deleteAt i extra)
+  case compare m n of
+    EQ -> pure base
+    GT -> errorMess ("The size(" ++ show m ++ ") of the 'base' set exceeds the target size(" ++ show n ++ ")") msgs
+    LT -> loop base possible
```
Slight simplification. The `EQ` case is taken care of by the first case in the `loop`
```suggestion
  if m > n
  then errorMess ("The size(" ++ show m ++ ") of the 'base' set exceeds the target size(" ++ show n ++ ")") msgs
  else loop base possible
```

TimSheard replied:
Don't agree. The surety that we cover all cases, trumps this. Left it as it is.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Combinators.hs:
```diff
@@ -0,0 +1,206 @@
+
+setSized :: Ord a => [String] -> Int -> Gen a -> Gen (Set a)
+setSized mess size gen = do
+  set <- (Set.fromList <$> vectorOf size gen)
```
```suggestion
  set <- Set.fromList <$> vectorOf size gen
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Combinators.hs:
```diff
@@ -0,0 +1,206 @@
+      | trys <= 0 =
+          if Set.size set == size
+            then pure set
+            else errorMess ("Ran out of trys(" ++ show numtrys ++ ") in fixSet: need " ++ show size ++ " elements, have " ++ show (Set.size set)) mess
```
Slight simplification. Redundant `if` (note the important change from `<=` to `<`):
```suggestion
      | trys < 0 =
          errorMess ("Ran out of tries(" ++ show numtrys ++ ") in fixSet: need " ++ show size ++ " elements, have " ++ show (Set.size set)) mess
```
I fixed the spelling in the error message, the binding however is up to you if you wanna fix it or not. This spelling is starting to grow on me :smile:

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Combinators.hs:
```diff
@@ -0,0 +1,206 @@
+            else errorMess ("Ran out of trys(" ++ show numtrys ++ ") in fixSet: need " ++ show size ++ " elements, have " ++ show (Set.size set)) mess
+    help trys set = case compare (Set.size set) size of
+      EQ -> pure set
+      GT -> help (trys - 1) (iterate Set.deleteMin set !! (Set.size set - size))
```
Here we just removing smallest values. Would it make sense to remove random values instead?

In any case, `!!` has a O(n) complexity. This is a bit better (requires `Data.Function (fix)` import):
```suggestion
      GT -> help (trys - 1) (fix (\f s n -> if n <= 0 then s else f (Set.deleteMin s) (n-1)) set (Set.size set - size))
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Combinators.hs:
```diff
@@ -0,0 +1,206 @@
+      LT -> do
+        let need = size - Set.size set
+        new <- Set.fromList . take need . filter (`Set.notMember` set) <$> vectorOf size genA
+        sized $ \n -> resize (n + 5) $ help (trys - 1) (Set.union new set)
```
There is a helper for this pattern:
```suggestion
        scale (+ 5) $ help (trys - 1) (Set.union new set)
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Combinators.hs:
```diff
@@ -0,0 +1,206 @@
+  fixSet (("setSized " ++ show size) : mess) 1000 size gen set
+
+fixSet :: Ord a => [String] -> Int -> Int -> Gen a -> Set a -> Gen (Set a)
+fixSet mess numtrys size genA s = help numtrys s
```
Honestly, this generator is implemented in a convoluted and inefficient way. Here is a clearer and more efficient way that I would write it:
```haskell
fixSet mess numTries size source
  | sz  > size = subsetFromSetWithSize ("fixSet":mess) set size
  | otherwise = = go numTries source
 where
    go n !set genA
      | n < 0 =
        errorMess ("Ran out of tries(" ++ show n ++ ") in fixSet: need " ++ show size ++ " elements, have " ++ show sz) mess
      | otherwise = do
        x <- genA
        if x `Set.member` set
          then go (n - 1) set $ scale (+1) genA
          else go n (Set.insert x set) genA
      where
        sz = Set.size set
```

This would probably need either a higher number of tries, just because each attempt is at generating one value, rather than a whole list.

TimSheard replied:
Left it the way it is.  I didn't write this. And don't wan to be responsible for breaking something that works.

TimSheard replied:
OK I used your code with a slight revision (scaling by 5 rather than 1). But I left the old code in a comment because when this breaks, I wan't to compare it with the previous version.

lehins replied:
Sounds good. The only reason why I changed to `(+1)` instead of `(+5)`, because it will be called a lot more often, then in the previous implementation. Reason for this is generation of each individual element one at a time, rather than in bulk using lists. Because it will be called more often it will scale up quicker too. At the end though, size is empirically discovered value, so it can be changed later if there are problems.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Combinators.hs:
```diff
@@ -0,0 +1,206 @@
+
+itemFromSet :: [String] -> Set a -> Gen a
+itemFromSet mess set | Set.null set = errorMess "itemFromSet : Can't take an item from the empty set." mess
+itemFromSet _ set = elements (Set.toList set)
```
This will be much faster:
```suggestion
itemFromSet _ set = \ix -> Set.elemAt ix set <$> chooseInt (0, Set.size set - 1)
```

FYI. Adding this function instead I think would even be more useful

```haskelll
-- | Remove a random element from a set
itemFromSet :: [String] -> Set a -> Gen (a, Set a)
itemFromSet mess set | Set.null set = ....
itemFromSet _ set = \ix -> (Set.elemAt ix set, Set.deleteAt ix set) <$> chooseInt (0, Set.size set - 1)
```

TimSheard replied:
done (with some parentheses I insist on)
 (\ix -> Set.elemAt ix set) <$> chooseInt (0, Set.size set - 1)
perhaps we could use a section?

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Combinators.hs:
```diff
@@ -0,0 +1,206 @@
+itemFromSet _ set = elements (Set.toList set)
+
+mapFromRange :: forall a b. Ord a => [b] -> Gen a -> Gen (Map a b)
+mapFromRange bs genA = Map.fromList <$> mapM (\b -> do a <- genA; pure (a, b)) bs
```
This generator is faulty. It will remove values from the range because there is no guarantee that generated keys will be unique. This is what we want instead:
```suggestion
mapFromRange bs genA = do
  keys <- setSized mess (length bs) genA
  pure $ mapFromDomRange keys bs
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Combinators.hs:
```diff
@@ -0,0 +1,206 @@
+mapFromRange bs genA = Map.fromList <$> mapM (\b -> do a <- genA; pure (a, b)) bs
+
+mapFromProj :: Ord a => [b] -> Gen a -> (b -> Gen c) -> Gen (Map a c)
+mapFromProj bs genA genC = Map.fromList <$> mapM (\b -> do a <- genA; c <- genC b; pure (a, c)) bs
```
Same problem here. Fix is even easier:
```suggestion
mapFromProj bs genA genC = do
  cs <- mapM genC bs
  mapFromRange cs genA
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Combinators.hs:
```diff
@@ -0,0 +1,206 @@
+mapFromProj bs genA genC = Map.fromList <$> mapM (\b -> do a <- genA; c <- genC b; pure (a, c)) bs
+
+mapFromDomRange :: Ord a => Set a -> [b] -> Map a b
+mapFromDomRange dom bs = Map.fromList $ zip (Set.toList dom) bs
```
Safe performance improvement:
```suggestion
mapFromDomRange dom bs = Map.fromAscList $ zip (Set.toAscList dom) bs
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Combinators.hs:
```diff
@@ -0,0 +1,206 @@
+mapSized mess size genA genB = setSized (("mapSized " ++ show size) : mess) size genA >>= addRange
+  where
+    addRange set = Set.foldl' accum (pure Map.empty) set
+    accum ansGen dom = do ans <- ansGen; rng <- genB; pure (Map.insert dom rng ans)
```
```suggestion
mapSized mess size genA genB = do
  keys <- setSized (("mapSized " ++ show size) : mess) size genA
  values <- vectorOf size genB
  mapFromDomRange keys values
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Combinators.hs:
```diff
@@ -0,0 +1,206 @@
+setFromSubsetWithSize :: Ord a => [String] -> Set a -> Int -> Gen a -> Gen (Set a)
+setFromSubsetWithSize mess subset n gen = do
+  additions <- setSized (("setFromSubsetWithSize " ++ show n) : mess) n gen
+  pure (Set.union subset additions)
```
`additions` could turn out to contain many elements that are already in `subset`, to a point that `Set.union subset additions == subset`. This might not be a big deal, considering that any set is its own subset. But this might be better:
```suggestion
  fixSet (("setFromSubsetWithSize " ++ show n) : mess) (n * 10) (n + Set.size subset) gen subset
```

TimSheard replied:
function was never used any where So I deleted it.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Combinators.hs:
```diff
@@ -0,0 +1,206 @@
+    help current count | count == n = pure current
+    help current count = do
+      index <- chooseInt (0, Set.size current - 1)
+      help (Set.deleteAt index current) (count - 1)
```
It took me forever to understand why there are two implementations of the same function. It would be nice to have some haddock for all of the generators, because it takes long time to figure out what those functions actually do 

If I understand correctly `subsetFromSetWithSize2` is more efficient when subset is big and `subsetFromSetWithSize` is more efficient when the majority of the elements are being removed. Right? I could not figure out any other reason why there would be two different implementations of the same generator be next to each other. So it's gotta be it. If that's the case then I would suggest one function that selects the most efficient implementation, which is easily determined by the number of elements that is needed to be removed or kept. This requires my suggestion to `itemFromSet`:

```suggestion
subsetFromSetWithSize :: Ord a => [String] -> Set a -> Int -> Gen (Set a)
subsetFromSetWithSize mess set n
  | Set.size set < n =
      errorMess
        ( "subsetFromSetWithSize: Can't make a subset of size "
            ++ show n
            ++ " from a smaller set of size "
            ++ show (Set.size set)
        )
        mess
   -- It is faster to remove extra elements, when we need to drop more than a half
  | Size.size `div` 2 < n = help fst (flip const) set Set.empty (Set.size - n)
   -- It is faster to pick out random elements when we only need under a half of the original set.
  | otherwise = help snd seq set Set.empty n
  where
    help pick optSeq !source target count
      | count <= 0 = pure $ pick (source, target)
      | otherwise = do
          (item, source') <- itemFromSet (("subsetFromSetWithSize " ++ show n) : mess) source
          -- To avoid a memory leak we only force the target when it is the set we want to keep.
          let target' = Set.insert item target
          target' `optSeq` help pick optSeq source' target' (count - 1)
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Combinators.hs:
```diff
@@ -0,0 +1,206 @@
+      ("from superSetFromSetWithSize" : msgs)
+      (size - Set.size setA)
+      (suchThat genA (`Set.notMember` setA))
+  pure (List.foldl' (flip Set.insert) setA additions)
```
Although elements in the generated `additions` are not going to be present in the `setA`, there is nothing preventing duplicates being generated in `additions` itself. This will produce a superset that is smaller then requested. We should use `fixSet` here instead. 

```suggestion
superSetFromSetWithSize msgs size genA setA = 
  let tries = 10 * (size - Set.size setA) -- at most 10 times per new element
  in fixSet ("superSetFromSetWithSize":mess) tries size setA genA
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Monad.hs:
```diff
@@ -0,0 +1,72 @@
+--   is constrained by the class 'c'. If one has an un-indexed type
+--   't' one can always use (Id t) instead. Eg
+--   With (Id x) <- hasOrd repT (Id t)
+data HasCond c t where
```
What does `Cond` stand for? I'd expect it to be called `HasConst`, for constraint.

TimSheard replied:
I renamed it to HasConstraint

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Rewrite.hs:
```diff
@@ -0,0 +1,307 @@
+        names = map getname xs
+        usesNames = map fst $ filter (any (`elem` names) . snd) $ map (\pr -> (pr, varsOfPred Set.empty pr)) cs0
+        theCycle = map show (names ++ [head names])
+        message = "Cycle in dependencies: " ++ List.intercalate " <= " theCycle
```
`head` is banned in our codebase:
```suggestion
        theCycle = map show (names ++ [head names])
        message = case names of
          [] -> "Empty dependencies"
          n:_ -> "Cycle in dependencies: " ++ List.intercalate " <= " map show (names ++ [n])
```

TimSheard replied:
Recall that there is a recursive cycle. So there must me at least two things in names.
How about
theCycle = case names of
            [] -> map show names
            (x:_) -> map show (names ++ [x])

lehins replied:
Sure that works.

> Recall that there is a recursive cycle. So there must me at least two things in names.

 I understand that it is impossible for a list to be empty in this case, but that should never be a reason to use a partial function like `head` that gives no clue where it has failed:
```
> head []
*** Exception: Prelude.head: empty list
```

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Rewrite.hs:
```diff
@@ -0,0 +1,307 @@
+    flatOrError (CyclicSCC xs : _) = failT [message, show info, unlines (map (("  " ++) . show) usesNames)]
+      where
+        names = map getname xs
+        usesNames = map fst $ filter (any (`elem` names) . snd) $ map (\pr -> (pr, varsOfPred Set.empty pr)) cs0
```
`elem` is O(n). This will be faster. Also list comprehension makes it bit clearer what is going on
```suggestion
        names = map getname xs
        namesSet = Set.fromList names
        usesNames = 
          [pr | pr <- cs0, any (`Set.member` namesSet) (varsOfPred Set.empty pr)]
```

TimSheard replied:
All this to print out an error message in a failled test. Don't think efficiency is an issue. But made the change anyway.

lehins replied:
Clarity was main motivator for the suggestion. I agree printing a message during testing doesn't need efficiency improvemnts, unless it is so slow that it makes the process hang. :slightly_smiling_face:

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Size.hs:
```diff
@@ -0,0 +1,302 @@
+data AddsSpec c where
+  AddsSpecSize ::
+    String ->
+    -- | name
```
Pipe `|` indicates haddock for the argument below, while `^` is for the argument above.
```suggestion
    -- ^ Name
```

TimSheard replied:
OK, got it.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Size.hs:
```diff
@@ -0,0 +1,302 @@
+    String ->
+    -- | name
+    Size ->
+    -- | total (range like (4 .. 12))
```
```suggestion
    -- ^ tTtal (range like (4 .. 12))
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Size.hs:
```diff
@@ -0,0 +1,302 @@
+  | SzAny
+  | SzLeast Int
+  | SzMost Int
+  | SzRng Int Int -- (SzRng i j) = [i .. j] . Invariant i <= j
```
Might as well turn comment into haddock
```suggestion
  | SzRng Int Int 
    -- ^ Size is in the range from @i@ to @j@ inclusive: @SzRng i j = [i .. j]@. Invariant @i <= j@
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Rewrite.hs:
```diff
@@ -0,0 +1,307 @@
+  mkDependGraph (length orderedNames) [] orderedNames [] simple
+
+showGraph :: (Vertex -> String) -> Graph -> String
+showGraph nameof g = unlines (map show (zip names (foldr (:) [] zs)))
```
```suggestion
showGraph nameof g = unlines (map show (zip names (toList zs)))
```

TimSheard replied:
But then I have to import Data.Foldable(toList)   done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Examples.hs:
```diff
@@ -0,0 +1,799 @@
+  pure (g y)
+
+stoi :: OrderInfo
+stoi = standardOrderInfo
```
In Russian "stoi" means "Don't move!" :smile:

TimSheard replied:
Ok I won't make a move.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Examples.hs:
```diff
@@ -0,0 +1,799 @@
+    Right x -> pure x
+  env <- monadTyped $ substToEnv subst emptyEnv
+  testTriples <- monadTyped (mapM (makeTest env) cs)
+  let messages2 = ("\nSubstitution produced after solving\n" : (map show subst))
```
```suggestion
  let messages2 = "\nSubstitution produced after solving\n" : map show subst
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Examples.hs:
```diff
@@ -0,0 +1,799 @@
+          then Nothing -- Indicates success, nothing bad happened
+          else Just ("Some conditions fail\n" ++ explainBad bad subst)
+  if loud
+    then trace (unlines (messages2 ++ messages3)) (pure $ ans)
```
```suggestion
    then trace (unlines (messages2 ++ messages3)) (pure ans)
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Examples.hs:
```diff
@@ -0,0 +1,799 @@
+-- | Test that 'cs' :: [Pred] does NOT have a solution. We expect a failure
+failn :: Proof era -> String -> Bool -> OrderInfo -> [Pred era] -> Assembler era -> IO ()
+failn proof message loud order cs target = do
+  putStrLn ("testing shouldFail test: " ++ message)
```
Tests should never print directly to `stdout`. Naturally it is acceptable during debugging (eg. with usage of `loud`), but not otherwise. It doesn't work well with test suites. Tests even can be executed in parallel and this could result in interleaved output.
```suggestion
  when loud $ putStrLn ("testing shouldFail test: " ++ message)
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Examples.hs:
```diff
@@ -0,0 +1,799 @@
+  shouldThrow
+    ( do
+        result <- generate (genMaybeCounterExample proof message loud order cs target)
+        case result of
+          Nothing -> putStrLn (message ++ " should have failed but it did not.")
+          Just counter ->
+            if loud
+              then putStrLn counter
+              else pure ()
```
```suggestion
        when loud $ case result of
          Nothing -> putStrLn (message ++ " should have failed but it did not.")
          Just counter -> putStrLn counter
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Examples.hs:
```diff
@@ -0,0 +1,799 @@
+              else pure ()
+    )
+    ( \(ErrorCall msg) ->
+        trace
+          ( if loud
+              then ("Fails as expected\n" ++ msg ++ "\nOK")
+              else ("Fails as expected OK")
+          )
+          True
```
```suggestion
        True <$ when loud (trace ("Fails as expected\n" ++ msg ++ "\nOK") (pure ()))
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Examples.hs:
This `Examples` module will need a lot of fixing. It's ok to put it off to a subsequent PR, but there are many things wrong with it.

TimSheard replied:
ok

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Solver.hs:
```diff
@@ -0,0 +1,670 @@
+testHasCond (ListR rep) t = case hasOrd rep [t] of
+  Typed (Right (With x)) -> print (synopsis (ListR rep) x)
+  Typed (Left xs) -> putStrLn $ unlines xs
+testHasCond _ _ = error "testHasCond only works on lists"
```
Unused function. Is needed for debugging? If not, removing it would be good.
```suggestion
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Solver.hs:
```diff
@@ -0,0 +1,670 @@
+-- ==================================
+
+-- | Is there an Count instance for 't'
+hasCount :: Rep era t -> (s t) -> Typed (HasCond Count (s t))
```
```suggestion
hasCount :: Rep era t -> s t -> Typed (HasCond Count (s t))
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Solver.hs:
```diff
@@ -0,0 +1,670 @@
+  pairs <- mapM (\d -> do r <- genRep repRng; pure (d, r)) ds
+  pure (Map.fromList pairs)
+  where
+    genThenOverwriteA a = do b <- genRep repDom; pure (Lens.set lensDomA a b)
```
```suggestion
    genThenOverwriteA a = Lens.set lensDomA a <$> genRep repDom
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Solver.hs:
```diff
@@ -0,0 +1,670 @@
+  where
+    accum spec cond = do
+      condspec <- (solveMap v cond)
+      (liftT (spec <> condspec))
```
```suggestion
      condspec <- solveMap v cond
      liftT (spec <> condspec)
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Solver.hs:
```diff
@@ -0,0 +1,670 @@
+  where
+    accum spec cond = do
+      condspec <- (solveSet v cond)
+      (liftT (spec <> condspec))
```
```suggestion
      condspec <- solveSet v cond
      liftT (spec <> condspec)
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Solver.hs:
```diff
@@ -0,0 +1,670 @@
+    zs = ("Cannot solve: genCount " ++ show v ++ " at type " ++ show r ++ " on Predicates") : map show cs
+
+-- | Used in solving Projections
+data Update t where Update :: Eq s => s -> Lens' t s -> Update t
```
```suggestion
data Update t where
  Update :: Eq s => s -> Lens' t s -> Update t
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Solver.hs:
```diff
@@ -0,0 +1,670 @@
+      | isAddsType r1 -> do
+          With v2 <- hasAdds r1 v1
+          sumv <- solveSums v2 cs
+          let msgs = (("Solving for variable " ++ nam) : map show preds)
```
```suggestion
          let msgs = ("Solving for variable " ++ nam) : map show preds
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Solver.hs:
```diff
@@ -0,0 +1,670 @@
+
+toolChain :: Era era => Proof era -> OrderInfo -> [Pred era] -> Gen (Env era)
+toolChain _proof order cs = do
+  (DependGraph pairs) <- monadTyped $ compile order cs
```
```suggestion
  DependGraph pairs <- monadTyped $ compile order cs
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs:
```diff
@@ -0,0 +1,1925 @@
+testMergeSize = do
+  spec1 <- genSize
+  spec2 <- genSize
+  case (spec1 <> spec2) of
```
```suggestion
  case spec1 <> spec2 of
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs:
```diff
@@ -0,0 +1,1925 @@
+  Some rep <- genAddsRep
+  spec1 <- genSizeByRep rep
+  spec2 <- genSizeByRep rep
+  case (spec1 <> spec2) of
```
```suggestion
  case spec1 <> spec2 of
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs:
```diff
@@ -0,0 +1,1925 @@
+relDisjoint r set = RelOper r Set.empty Nothing set
+relEqual r set = RelOper r set (Just set) Set.empty
+
+instance Monoid (RelSpec era dom) where mempty = RelAny
```
```suggestion
instance Monoid (RelSpec era dom) where
  mempty = RelAny
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs:
```diff
@@ -0,0 +1,1925 @@
+instance Semigroup (RelSpec era dom) where
+  (<>) = mergeRelSpec
+
+instance Show (RelSpec era dom) where show = showRelSpec
```
```suggestion
instance Show (RelSpec era dom) where
  show = showRelSpec
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs:
```diff
@@ -0,0 +1,1925 @@
+
+okSize :: RelSpec era d -> Bool
+okSize (RelOper _ must (Just may) cant) =
+  (Set.size must) <= (Set.size (Set.difference may cant))
```
```suggestion
  Set.size must <= Set.size (Set.difference may cant)
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs:
```diff
@@ -0,0 +1,1925 @@
+-- | return a generator that always generates things that meet the RelSpec
+genFromRelSpec :: forall era t. Ord t => [String] -> Gen t -> Int -> RelSpec era t -> Gen (Set t)
+genFromRelSpec msgs g n spec =
+  let msg = ("genFromRelSpec " ++ show n ++ " " ++ show spec)
```
```suggestion
  let msg = "genFromRelSpec " ++ show n ++ " " ++ show spec
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs:
```diff
@@ -0,0 +1,1925 @@
+--   Note that the empty set is always a solution.
+--   These sets tend to be rather small (size <= atleastdelta)
+genDisjoint :: Ord a => Set a -> Gen a -> Gen (Set a)
+genDisjoint set gen = help atleastdelta Set.empty
```
It would be nice to use camelCase. It hurts the eye to read function names like that.
```suggestion
genDisjoint set gen = help atLeastDelta Set.empty
```

TimSheard replied:
done (and 6 other substitutions)

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs:
```diff
@@ -0,0 +1,1925 @@
+genDisjoint set gen = help atleastdelta Set.empty
+  where
+    help n answer | n < 0 = pure answer
+    help n answer = do
```
```suggestion
    help n !answer | n < 0 = pure answer
    help n !answer = do
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs:
```diff
@@ -0,0 +1,1925 @@
+  s2 <- genConsistentRelSpec ["testConsistentRel " ++ show n ++ " " ++ show s1] (choose (1, 1000)) s1
+  case s1 <> s2 of
+    RelNever ms -> pure $ counterexample (unlines (["genConsistent fails", show s1, show s2] ++ ms)) False
+    _ -> pure $ counterexample "" True
```
```suggestion
    _ -> pure $ property True
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs:
```diff
@@ -0,0 +1,1925 @@
+  s1 <- genRelSpec (("genRelSpec") : msg) (choose (1, 10000)) Word64R n
+  s2 <- genConsistentRelSpec (("genConsistentRepSpec " ++ show s1) : msg) (choose (1, 1000)) s1
+  let s3 = (s1 <> s2)
+  case s3 of
```
`s3` isn't used anywhere else
```suggestion
  case s1 <> s2 of
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs:
```diff
@@ -0,0 +1,1925 @@
+  s2 <- genConsistentRelSpec (("genConsistentRepSpec " ++ show s1) : msg) (choose (1, 1000)) s1
+  let s3 = (s1 <> s2)
+  case s3 of
+    RelNever _ -> (pure $ property True) -- This test is an implication (consistent (s1 <> s2) => ...)
```
```suggestion
    RelNever _ -> pure (property True) -- This test is an implication (consistent (s1 <> s2) => ...)
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs:
```diff
@@ -0,0 +1,1925 @@
+  -- | Something was inconsistent
+  RngNever :: [String] -> RngSpec era rng
+
+instance Show (RngSpec era t) where show = showRngSpec
```
```suggestion
instance Show (RngSpec era t) where
  show = showRngSpec
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs:
```diff
@@ -0,0 +1,1925 @@
+
+instance Show (RngSpec era t) where show = showRngSpec
+
+instance Monoid (RngSpec era rng) where mempty = RngAny
```
```suggestion
instance Monoid (RngSpec era rng) where
  mempty = RngAny
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs:
```diff
@@ -0,0 +1,1925 @@
+      pr x@(_, a, _, b, _, c, _) = if not (a && b && c) then Just (showAns x) else Nothing
+  let trips = [(x, y, m) | x <- xs, y <- ys, Just m <- [consistent x y]]
+  ts <- mapM check trips
+  pure $ (n, length trips, Maybe.catMaybes (map pr ts))
```
```suggestion
  pure $ (n, length trips, Maybe.mapMaybe pr ts)
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs:
```diff
@@ -0,0 +1,1925 @@
+  -- | Something is inconsistent
+  MapNever :: [String] -> MapSpec era dom rng
+
+instance Ord d => Show (MapSpec w d r) where show = showMapSpec
```
```suggestion
instance Ord d => Show (MapSpec w d r) where
  show = showMapSpec
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs:
```diff
@@ -0,0 +1,1925 @@
+
+instance Ord d => Show (MapSpec w d r) where show = showMapSpec
+
+instance (Ord dom) => Semigroup (MapSpec era dom rng) where (<>) = mergeMapSpec
```
```suggestion
instance Ord dom => Semigroup (MapSpec era dom rng) where
  (<>) = mergeMapSpec
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs:
```diff
@@ -0,0 +1,1925 @@
+
+instance (Ord dom) => Semigroup (MapSpec era dom rng) where (<>) = mergeMapSpec
+
+instance (Ord dom) => Monoid (MapSpec era dom rng) where mempty = MapSpec SzAny RelAny RngAny
```
```suggestion
instance Ord dom => Monoid (MapSpec era dom rng) where
  mempty = MapSpec SzAny RelAny RngAny
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs:
```diff
@@ -0,0 +1,1925 @@
+--   inconsistencies there as well as (\ a b c -> dropT (mapSpec a b c)) has the same
+--   type as MapSpec, but pushes the reports of inconsistencies into MapNever.
+mapSpec :: Ord dom => Size -> RelSpec era dom -> RngSpec era rng -> Typed (MapSpec era dom rng)
+mapSpec sz1 rel rng = case (sz1 <> sz2 <> sz3) of
```
```suggestion
mapSpec sz1 rel rng = case sz1 <> sz2 <> sz3 of
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs:
```diff
@@ -0,0 +1,1925 @@
+    then pure (Map.fromList (zip (Set.toList dom) rangelist))
+    else
+      trace ("Dsize=" ++ show (Set.size dom) ++ " =/=  Rsize=" ++ show (length rangelist)) $
+        (pure (Map.fromList (zip (Set.toList dom) rangelist)))
```
Tracing shouldn't be used in tests, especially in generators
```suggestion
  pure (Map.fromList (zip (Set.toList dom) rangelist))
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs:
```diff
@@ -0,0 +1,1925 @@
+    msgs = ["genSetSpecIsSound"]
+
+manyMergeSetSpec :: Gen (Int, Int, [String])
+manyMergeSetSpec = do
```
It would be nice to reduce some duplication. I've identified 3 nearly identical functions. I am sure there are more

TimSheard replied:
These functions are similar but different. There are some notes in comments at the bottom of this file about introducing a class, so they all could be a method of such a class. The problem is that some of the parts take different numbers and types of arguments. So it would have to be a class with multiple associated type families. There is some hope, but look at the notes and you will see how complicated it gets.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs:
```diff
@@ -0,0 +1,1925 @@
+  ElemNever :: [String] -> ElemSpec era t
+  ElemAny :: ElemSpec era t
+
+instance Show (ElemSpec era a) where show = showElemSpec
```
```suggestion
instance Show (ElemSpec era a) where
  show = showElemSpec
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs:
```diff
@@ -0,0 +1,1925 @@
+    ElemSpec era x
+  -- | The range has exactly these elements
+  ElemEqual :: Eq t => Rep era t -> [t] -> ElemSpec era t
+  -- ElemOrd :: Ord t => Rep era t -> t -> OrdCond -> t -> ElemSpec era t
```
Dead code, or placeholder for the future?
```suggestion
```

TimSheard replied:
Something we want to add in the future, left a comment

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs:
```diff
@@ -0,0 +1,1925 @@
+  ListSpec :: Size -> ElemSpec era t -> ListSpec era t
+  ListNever :: [String] -> ListSpec era t
+
+instance Show (ListSpec era a) where show = showListSpec
```
```suggestion
instance Show (ListSpec era a) where
  show = showListSpec
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Tests.hs:
```diff
@@ -0,0 +1,680 @@
+-- Orphan Arbitrary instance for OrderInfo
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+-- {-# OPTIONS_GHC -Wname-shadowing #-}
```
```suggestion
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Tests.hs:
```diff
@@ -0,0 +1,680 @@
+genMapLiteralWithDom env vrep keys = do
+  let genVal = do
+        v <- genLiteral env vrep
+        pure v
```
```suggestion
  let genVal = genLiteral env vrep
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Tests.hs:
```diff
@@ -0,0 +1,680 @@
+genKeyType = elements [TypeInEra IntR]
+
+genValType :: Gen (TypeInEra era)
+genValType = elements [TypeInEra CoinR, TypeInEra DeltaCoinR] -- , TypeInEra (PairR CoinR CoinR)]
```
Forgotten or needed for future?
```suggestion
genValType = elements [TypeInEra CoinR, TypeInEra DeltaCoinR]
```

TimSheard replied:
I didn't writ this code, but I think this is a reminder that this may be needed in the future. I added a comment to say that.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Tests.hs:
```diff
@@ -0,0 +1,680 @@
+
+-- | Unsatisfiable constraint returned if we fail during constraint generation.
+errPred :: [String] -> Pred era
+errPred errs = (Lit (ListR StringR) ["Errors:"]) :=: (Lit (ListR StringR) errs)
```
```suggestion
errPred errs = Lit (ListR StringR) ["Errors:"] :=: Lit (ListR StringR) errs
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Tests.hs:
```diff
@@ -0,0 +1,680 @@
+shrinkPreds :: ([Pred era], GenEnv era) -> [([Pred era], GenEnv era)]
+shrinkPreds (preds, env) =
+  [ (preds', env') | preds' <- shrinkList shrinkPred preds, let defined = foldMap def preds'
+                                                                env' = env {gEnv = pruneEnv defined (gEnv env)}, depCheck mempty preds'
```
Current formatting is unreadable. It could be even cleaner to switch to `do` syntax for lists, instead of list comprehensions.
```suggestion
  [ (preds', env') | preds' <- shrinkList shrinkPred preds
  , let defined = foldMap def preds'
        env' = env {gEnv = pruneEnv defined (gEnv env)}
  , depCheck mempty preds'
```

TimSheard replied:
Didn't write this, But I agree this needed reformatting.

lehins reviewed (approved):
Bunch of syntactic suggestions. However there are a few useful comments and suggestions about generators.

Massive piece of work!
